### PR TITLE
Cache: Simple http cache merge header entries with rules

### PR DESCRIPTION
**Pull request by tangsaidi**
Commit Message: When validating, we merge the incoming headers into our existing cache. We skip headers that we don't want to include.
Risk Level: Low
Testing: Unit, Integration
Fixes: existing TODO
Signed-off-by: tangsaidi@google.com

### CODE REVIEW

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (dmitri-d) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:131]: Would this be better implemented as a const member?

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (dmitri-d) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:377]: Could you add a unit-level test that calls `SimpleHttpCache::updateHeaders` directly and verifies that every field in `headersNotToUpdate` remains unchanged?

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (capoferro) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:40]: Can this just be a `LookupContext&` instead of a reference to a unique_ptr? We dereference it on hte next line, anyway, so perhaps we just do that at the call sites for this `updateHeaders`.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (capoferro) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:160]: perhaps we should say "not handled yet" as we wouldn't be against someone adding support for them, but we just don't care to spend time on it right now.

**Comment by RyanTheOptimist** [comment]
HTTP Caching is complex. :) The [most recent IETF doc on the subject](https://www.ietf.org/archive/id/draft-ietf-httpbis-cache-18.html) currently in Last Call at the HTTP Working group. I think it would be nice to understand how our caching does or doesn't match up with the semantics in that doc. In particular, how did we decide which headers belong in `headersNotToUpdate` and which do not. Is there an open issue which motivated this change?

**Comment by tangsaidi** [comment]
@RyanTheOptimist Hi Ryan, the change is referencing the previous [RFC7234](https://datatracker.ietf.org/doc/html/rfc7234). It looks like they made some enhancement to the specs in the latest IETF docs. I'll read and amend the PR based on that. Most of the `headersNotToUpdate` list is based on [these principles](https://www.ietf.org/archive/id/draft-ietf-httpbis-cache-18.html#name-updating-stored-header-fiel). I'll make sure to add some explanations.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (tangsaidi) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:160]: It looks like according to the latest [IETF doc](https://www.ietf.org/archive/id/draft-ietf-httpbis-cache-18.html) recognizes the warning codes are obsolete and removed it from the specs. I'll probably remove this note altogether.

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (tangsaidi) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc]: I've added a unit test that checks for every field except for `vary`. It's infeasible to test `vary` because we are skipping `update_header` operation as long as it has the vary field. We'll be implementing the `vary` logic soon.

**Comment by jmarantz** [comment]
I have a higher level question about the general direction indicated by this PR.

Say someone wants to write an in-memory LRU cache for Envoy. Can the SimpleHttpCache implementation here be used as a base class, so that all the RFC semantics encapsulated here can be re-used?

@toddmgreer @capoferro

**Comment by capoferro** [comment]
> I have a higher level question about the general direction indicated by this PR.
> 
> Say someone wants to write an in-memory LRU cache for Envoy. Can the SimpleHttpCache implementation here be used as a base class, so that all the RFC semantics encapsulated here can be re-used?

The goal of the CachePolicy (which I have a PR prepared to submit once my other PRs are merged) is to provide that sort of extension point. HttpCache implementations should not contain cache semantics after the CachePolicy exists. Once I complete the base implementation of CachePolicy, we should add this to the config there, I think.

**Comment by toddmgreer** [comment]
> I have a higher level question about the general direction indicated by this PR.
> 
> Say someone wants to write an in-memory LRU cache for Envoy. Can the SimpleHttpCache implementation here be used as a base class, so that all the RFC semantics encapsulated here can be re-used?

SimpleHttpCache isn't designed to be a base class, so I would be skeptical of deriving from it. As capoferro said, CachePolicy is intended to handle RFC semantics. There may be some niche cache implementations that find it convenient to handle some oddball header logic, but they should mostly be as close to simple caches as possible.

**Comment by tangsaidi** [comment]
> I have a higher level question about the general direction indicated by this PR.
> 
> Say someone wants to write an in-memory LRU cache for Envoy. Can the SimpleHttpCache implementation here be used as a base class, so that all the RFC semantics encapsulated here can be re-used?
> 
> @toddmgreer @capoferro

I think the `SimpleHttpCache` serves as a simple example on how to extend the `HttpCache`. As @capoferro  suggested, we probably will need to move the RFC semantics within this PR over to `CachePolicy` once its ready.

**Comment by jmarantz** [comment]
Directionally this seems fine. Once @RyanTheOptimist approves it I'll do a review round.

**Review by RyanTheOptimist — DISMISSED** [review]
LGTM

Thanks for the additional context and comments!

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:113]: use CONSTRUCT_ON_FIRST_USE pattern here to avoid an init-order fiasco.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:166]: I'm assuming (though it's not 100% clear from context) that you are modifying a different set of headers than the one you are scanning. Is that right? Otherwise I think response_header.value() might not be valid after the remove. Maybe just comment about that?

In any case, rather than 'remove' and then 'addCopy' on the same header, can you just find the HeaderEntry and then set it? I think that would make the code a little more complex, but faster. It'll be more complex because you'd have to special-case when there was more than one header of the same name (which is likely less common).

See HeaderMap::get which returns a GetResult, and you can just set the value of that, if there's only one of them. Then you can do the remove/addCopy dance or something else only if there was more than one.

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:415]: put this formatted time in a temp rather than calling fromTime multiple times. In particular, it is not obvious whether current_time_ might advance between the calls, and temping it would put that issue to rest.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:166]: Also I'm not sure if it is valid HTTP to duplicate any of the headers in your set. Of course you need to handle it, but it makes sense to optimize for the single-entry only case.

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:429]: can you cover a case where the input headers you are processing have a case where a header appears more than once?

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (tangsaidi) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:415]: Thanks for pointing this out!
```
  std::string str_1 = formatter_.fromTime(current_time_);
  time_source_.advanceTimeWait(Seconds(3601));
  std::string str_2 = formatter_.fromTime(current_time_);
```
This is a pattern that's used a lot in this test file. It's an incorrect use of the `time_source_` and `current_time_`. The `current_time_` doesn't get updated upon `time_source_.advanceTimeWait()`.  We only found out about this yesterday. I'll fix all occurrences of this pattern in this PR.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (toddmgreer) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:162]: This check should be moved to CacheFilter::shouldUpdateCachedEntry, which should return false if a 304 contains a Content-Length field (unless it's equal to the current one). I don't regard that as blocking this PR though, since it's a pre-existing issue.

**Review by toddmgreer — CHANGES_REQUESTED** [review]

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (tangsaidi) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:166]: > In any case, rather than 'remove' and then 'addCopy' on the same header, can you just find the HeaderEntry and then set it? I think that would make the code a little more complex, but faster. It'll be more complex because you'd have to special-case when there was more than one header of the same name (which is likely less common).

The `HeaderMap` doesn't have an API that returns a non-const reference to the entry of interest. Specifically, the `GetResult` class has the following documentation.
```
 /**
   * This is a wrapper for the return result from get(). It avoids a copy when translating from
   * non-const HeaderEntry to const HeaderEntry and only provides const access to the result.
   */
```

Adding this optimization means that we need to change the `header_map`'s API interface, which is used by lots of http filters. IMHO, the optimization is not worth the risk. It's still O(n) for the single-entry use case because we need to scan through the headermap to ensure there's no duplicate.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (tangsaidi) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:130]: Line 147 is only a placeholder for the `vary` logic. We probably won't return from there once the `vary` logic is fully implemented. I don't think it hurts to include it here, as I can't think of a scenario that it makes sense to update `vary` field.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (tangsaidi) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:162]: This basically handles individual keys that we do not want to update, given that we already decided we want to update the entry. I'm adding additional checks for these items in `shouldUpdateCachedEntry` in my next PR. I'll make sure to add you as a reviewer there.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:113]: leaving this unresolved until it's....resolved :)

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:166]: OK I think you're right; there's no good way to do the multi-value get. But can you what happens if you use set? That might be a single-call mechanism that can give you the same semantics.

Just want to make sure that if there are multiple values that they'd all be replaced by the one value you are specifying.

Comment (tangsaidi) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:166]: Thanks for pointing this out! It looks like `setCopy` does exactly remove and add. 
```

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:401]: nit: `const std::string time_now`

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (tangsaidi) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:165]: @jmarantz This `updatedHeaderFields` hash set handles the case where the incoming headers have multiple values for one field. A call to `setCopy` would overwrite previous values as we iterate through the response_headers. Another reason why I kept the `remove` and `addCopy` separate is because the behaviour of `setCopy` can be implementation dependent. In our current implementation, it is equivalent to `remove` then `add`. But [sometimes](https://github.com/envoyproxy/envoy/pull/9114/files#diff-2920212df144c38f64da08a4094a67b3c897669460307a6ed0147461fa2d4907R494), the `setCopy` would only replace the first occurrence.

Comment (jmarantz) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:165]: can you document that nuance in a comment here? Maybe also note that if setCopy is changed to remove all fields then this can be simplified.

Also as the fields in the set are well-known, does that multiple-value case ever occur for them?

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:446]: the case you have below is good; keep it. But I was thinking of a case where one of the headers in your update-list got repeated. Could you add one of those also?

**Review by jmarantz — COMMENTED** [review]
/wait

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (tangsaidi) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:165]: The link above directs to an old commit from 2 years ago. Our current `setCopy` implementation removes all fields, But we can't really simplify the code, because then we still need to use `addCopy` for special cases where the incoming headers have multiple-value. It's basically same amount of code
```   
if (updatedHeaderFields.contains(lower_case_key)) {
   addCopy(lower_case_key, value)
   updatedHeaderFields.insert(lower_case_key)
} else {
   setCopy(lower_case_key, value)
}
```
vs
```
if (!updatedHeaderFields.contains(lower_case_key)) {
   remove(lower_case_key)
   updatedHeaderFields.insert(lower_case_key)

}
   addCopy(lower_case_key, value)

```

And yes, the header_map may have multiple values for the same key because not all the headers are inlined.

Comment (tangsaidi) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:165]: I'll add a comment to explain this

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (tangsaidi) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:448]: In the original entry, I repeated `etag` twice. In the incoming headermap, I added `content-length` twice. Although neither makes sense in a real-world scenario, these two cases should make sure the code is behaving correctly even if there's some issues with the client.

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:168]: Suggest:
```
// The behavior of setCopy is different between built-in O(1) headers and arbitrary headers. With
// the latter, only the first occurrence is changed, which is not what we need here; we always want
// to replace all the occurrences. Thus we use `remove` and then `addCopy`.
```

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (jmarantz) [inline test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:452]: Sorry I'm actually confused now by your test structure. Could you rename response_headers_3 to expected_response_headers, and response_headers_2 to origin_response_headers, or similar? It's possible that response_headers_2 is actually cached_response_headers; not clear from looking at this code snippet.

Just numbering them doesn't show their

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (tangsaidi) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:168]: Sorry, I think I didn't explain the state of `setCopy` well. For both O(1) headers and arbitrary headers, all occurrences would be replaced by exactly one new value. But whether it actually does that is dependent on how `HeaderMapImpl` implements it. I think I'll add this part of the suggestion
```
// we always want to replace all the occurrences. Thus we use `remove` and then `addCopy`.
```

File: test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc
Comment (tangsaidi) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:452]: For sure! Sorry about the confusion

Comment (tangsaidi) [reply test/extensions/filters/http/cache/simple_http_cache/simple_http_cache_test.cc:453]: Fixed 😀

**Review by toddmgreer — DISMISSED** [review]

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:165]: Thanks for changing to the simple setCopy call. I didn't realize that the impl itself does 2 lookups for this one call; seems not necessary but that's a topic for a different PR.

Can you update this comment now?

**Review by jmarantz — COMMENTED** [review]
lgtm modulo comment tweak.

/assign-from @envoyproxy/senior-maintainers

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (tangsaidi) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: Done :D

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (jmarantz) [inline source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc:157]: typo?

**Review by jmarantz — COMMENTED** [review]
/wait

File: source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc
Comment (tangsaidi) [reply source/extensions/filters/http/cache/simple_http_cache/simple_http_cache.cc]: Good catch, thanks! Fixed

**Review by alyssawilk — APPROVED** [review]
LGTM for non-extension changes, trusting jmarantz on the rest unless he explicitly wants me to do a pass there

**Pull request merged**
